crates/compiler/can/src/expr.rs

```diff
@@ -729,6 +731,54 @@ pub fn canonicalize_expr<'a>(
 
         ast::Expr::Str(literal) => flatten_str_literal(env, var_store, scope, literal),
 
+        ast::Expr::IngestedFile(file_path, type_ann) => {
+            use ast::TypeAnnotation::Apply;
+
+            if let Apply("", "Str", &[]) = type_ann.value {
+                match std::fs::read_to_string(file_path) {
+                    Ok(data) => (Expr::Str(data.into_boxed_str()), Output::default()),
+                    // TODO: Also handle when it returns an error converting to UTF8 explicitly.
+                    Err(e) => todo!("Handle err case of loading a string from a file: {:?}", e),
+                }
+            } else if let Apply(
+                "",
+                "List",
+                &[Loc {
+                    value: Apply("", "U8", &[]),
+                    ..
+                }],
+            ) = type_ann.value
+            {
```

Comment:
I would structure this a bit differently:

- During canonicalization here, simply check if the file exists. If it doesn't, yield a canonicalization error. Otherwise, load the file's bytes, say to a new `Expr::IngestedFile`.
- Save the resolution of the type (as you've done with the `Apply` checks here) until type checking. There, we can check that the `Expr::IngestedFile` actually satisfies a Str, List U8, etc, and will give us a way that will be easier to extend later on. At this point, the `Expr::IngestedFile` will still be a list of bytes, but if we see for example it should check as a Str, we can also do the UTF-8 check at that point.  This will give us a way to extend IngestedFile checking to other types, including custom decodable types, more easily later on.
- Finally, in mono, actually do the conversion to a String or list of bytes as necessary after the type has been resolved.

All the other implementation including parsing code LGTM.

Reply:
Ok. I think I now have the feature properly working. That said, I need to deal with properly creating errors and handling the utf8 check. So I think with robust errors, this should be essentially correct.